pkg/workceptor/kubernetes.go

```diff
@@ -338,115 +343,174 @@ func (kw *KubeUnit) KubeLoggingWithReconnect(streamWait *sync.WaitGroup, stdout
 						podNamespace,
 						podName,
+						err.Error(),
 					)
 
 					return
 				}
 
-				podConditionReady := false
-				erroredPod, kubeErr := kw.KubeAPIWrapperInstance.Get(kw.GetContext(), kw.clientset, podNamespace, podName, metav1.GetOptions{})
+				// EOF errors are expected in two cases.
+				// 1. When the job is finished and the last line is sent.
+				// In this case we monitor the container status to ensure we move out of Running,
+				// and we make sure we get the last line of output.
+				// 2. When the job lasts longer than 4 hours and kube api closes the log stream.
+				// This is a recoverable EOF, so we attempt to reconnect with a back-off.
+				// In BOTH cases, we have a simular approach, wait 1-2 seconds and check if the container status has changed.
+
+				podDetails, kubeErr := kw.KubeAPIWrapperInstance.Get(kw.GetContext(), kw.clientset, podNamespace, podName, metav1.GetOptions{})
 				if kubeErr != nil {
-					kw.GetWorkceptor().nc.GetLogger().Debug("Error getting pod after reading stream: '%s'", kubeErr)
+					// Their are many reasons why the kube api might not be able to get the pod,
+					// This does not mean their is a problem just yet.
+					// Lets try to get the pod again, max 5 times, and decide.
+					kw.GetWorkceptor().nc.GetLogger().Debug("Error getting pod after reading stream: '%s' , continuing try to get pod up to 5 more times.", kubeErr)
+
+					continue mainLoop
 				}
-				for _, condition := range erroredPod.Status.Conditions {
-					if condition.Type == corev1.PodReady && condition.Status == corev1.ConditionTrue {
-						podConditionReady = true
+
+				var containerState corev1.ContainerState
+				foundContainer := false
+				for _, containerStatus := range podDetails.Status.ContainerStatuses {
+					if containerStatus.Name == WorkerContainerName {
+						containerState = containerStatus.State
+						foundContainer = true
 					}
 				}
 
-				if err == io.EOF && !podConditionReady {
+				if !foundContainer {
+					kw.GetWorkceptor().nc.GetLogger().Error("Unable to find the container %s for pod %s. This is unrecoverable. Marking the job as failed and exiting", WorkerContainerName, podName)
+					*stdoutErr = fmt.Errorf("unable to find the container %s for pod %s. This is unrecoverable. Marking the job as failed and exiting", WorkerContainerName, podName)
+
+					return
+				}
+
+				switch {
+				case containerState.Running != nil:
+					// We got EOF but pod is running, is this because we checked too fast? Will it turn into a terminated state soon or are we hitting the 4 hour log stream kube error. We will attempt to reconnect a max of 5 times in order to cover both cases
+					// If we can't get reconnect without an EOF we will error and mark the job as failed.
+					retryGetLogStream--
+					if retryGetLogStream > 0 {
+						kw.GetWorkceptor().nc.GetLogger().Info(
+							"Detected Error: %s for pod %s/%s. Will retry %d more times.",
```

Comment:
```suggestion
							"Detected EOF Error: %s for pod %s/%s in with container state: Running. Job may not be complete. Will retry %d more times.",
```

It will be really helpful for the log messages to explicitly report the state. Here we know we encountered an `EOF` error and we fetched the container but it is still in `Running` state.

What do you think of including the `curContainerDelay` (the time we're about to Sleep) in these log messages too? Just so we can see in the logs that delay?

Reply:
I'm not sure about adding the delay time in the log. We already have the count so we know where the sleep happen. 🤔

Reply:
👍